components/ws-manager-mk2/service/manager.go

```diff
@@ -234,11 +234,10 @@ func (wsm *WorkspaceManagerServer) StartWorkspace(ctx context.Context, req *wsma
 
 func (wsm *WorkspaceManagerServer) StopWorkspace(ctx context.Context, req *wsmanapi.StopWorkspaceRequest) (*wsmanapi.StopWorkspaceResponse, error) {
 	err := wsm.modifyWorkspace(ctx, req.Id, true, func(ws *workspacev1.Workspace) error {
-		ws.Status.Conditions = append(ws.Status.Conditions, metav1.Condition{
+		ws.Status.Conditions = wsk8s.AddUniqueCondition(ws.Status.Conditions, metav1.Condition{
```

Comment:
The values in the conditions array are meant to be unique (they're keyed) on their `Type`. This is just fixing the logic to match what we do elsewhere. If another stop request comes in we will update the `LastTransitionTime` of the condition

Reply:
>  This is just fixing the logic to match what we do elsewhere.

Can you share a link to where else we do this? I'm just curious to see where else.

> If another stop request comes in we will update the LastTransitionTime of the condition

I see, in other words, this refactor brings the logic which governs the update of Conditions from elsewhere to `wsk8s.AddUniqueCondition`.

Reply:
@WVerlaek for :eyes: on :point_up:

Reply:
@kylos101 see e.g. https://github.com/gitpod-io/gitpod/blob/468f7bf0166c04950ad783a42453138bbcac9780/components/ws-daemon/pkg/controller/controller.go#L188 or https://github.com/gitpod-io/gitpod/blob/468f7bf0166c04950ad783a42453138bbcac9780/components/ws-manager-mk2/controllers/status.go#L51

Basically, anywhere where we modify the workspace to set a condition, we should use `wsk8s.AddUniqueCondition` instead of appending it to the conditions array, to ensure that we only have one condition of each `Type`

Reply:
Fantastic, thank you very much for sharing, @WVerlaek ! :pray: TIL. :smile_cat: